packages/headless-react/src/hooks/controller-hook.ts

```diff
@@ -0,0 +1,159 @@
+  T,
+  'subscribe' | 'state'
+>;
+
+export interface UseControllerReturnType<T extends Controller> {
+  state: T['state'];
+  controller: ReactController<T>;
+}
+
+function bindMethods<T extends object>(methods: T, obj: object): T {
+  Object.entries(methods).forEach(([key, value]) => {
+    if (typeof value === 'function') {
+      (methods as Record<string, unknown>)[key] = value.bind(obj);
+    } else if (typeof value === 'object') {
+      bindMethods(value, (obj as Record<string, object>)[key]);
+    }
+  });
+  return methods;
+}
+
+function getOnStateChanged(subscribe: (listener: () => void) => Unsubscribe) {
+  return (listener: () => void) => {
+    let firstCall = true;
+    return subscribe(() => {
+      if (firstCall) {
+        firstCall = false;
+        return;
+      }
+      listener();
+    });
+  };
+}
+
+function buildController(
+  builder: (engine: CoreEngine, props?: object) => Controller,
+  engine: CoreEngine,
+  props?: object
+) {
+  const headlessController = builder(engine, props);
+  const {subscribe: _, state: __, ...methods} = headlessController;
+  return {
+    controller: bindMethods(methods, headlessController),
+    getStateSnapshot: () => structuredClone(headlessController.state),
+    onStateChanged: getOnStateChanged((listener) =>
+      headlessController.subscribe(listener)
+    ),
+  };
+}
+
+/**
+ * Same as [useSyncExternalStore](https://reactjs.org/docs/hooks-reference.html#usesyncexternalstore), but doesn't infinitely re-render with a memoized store.
+ */
+function useSyncMemoizedStore<T>(
+  subscribe: (listener: () => void) => () => void,
+  getSnapshot: () => T
+) {
+  const state = useRef<T | null>(null);
+  const [, render] = useReducer((renderCount: number) => renderCount + 1, 0);
+  useEffect(
+    () =>
+      subscribe(() => {
+        state.current = getSnapshot();
+        render();
+      }),
+    [subscribe, getSnapshot]
+  );
+  if (useHasChanged([subscribe, getSnapshot]) || state.current === null) {
+    state.current = getSnapshot();
+  }
+  return state.current as T;
+}
+
+export function useController<
+  TController extends Controller,
+  TEngine extends CoreEngine
+>(
+  builder: (engine: TEngine) => TController,
+  engine: TEngine
+): UseControllerReturnType<TController>;
+export function useController<
+  TController extends Controller,
+  TEngine extends CoreEngine,
+  TProps extends object
+>(
+  builder: (engine: TEngine, props: TProps) => TController,
+  engine: TEngine,
+  props: TProps
+): UseControllerReturnType<TController>;
+export function useController(
+  builder: (engine: CoreEngine, props?: object) => Controller,
+  engine: CoreEngine,
+  props?: object
+): UseControllerReturnType<Controller> {
+  if (useHasChanged([builder])) {
+    console.error(
+      'A Coveo Headless controller builder was changed in between renders. This can happen when a custom callback is passed as a builder instead of an build function imported directly from Headless.'
+    );
+  }
```

Comment:
If the user initializes a controller like so, React will detect `builder` as being changed on every single re-render:
```
const { controller, state } = useController((e, p) => buildFacet(e, p), engine, props);
```

This is because React uses `Object.is(a, b)` to compare whether a value in a dependency array was changed, which essentially compares pointers as opposed to a deep comparison, and callbacks are passed as-is to functions.

I made the `useOnChange` hook so that we could log an error when the reference is changed. It works almost exactly like `useEffect`, except:
* It's called synchronously on render as opposed to after the render.
* It's only called when the value is changed, never on the initial value.

Reply:
edit: Ok, I saw in your description that this is already something you've thought of as a possible improvement.

Great !

Reply:
Yes! I agree that this would be more idiomatic. I also just thought of another advantage: we would know at build-time whether one of our controllers doesn't match the expected parameters.

I will go with this approach then.